noise.zig

```diff
@@ -0,0 +1,668 @@
+    }
+
+    pub fn split(self: SymmetricState) [2]CipherState {
+        var tmp_k1 = EMPTY_HASH;
+        var tmp_k2 = EMPTY_HASH;
+        var out2 = EMPTY_HASH;
+
+        hkdf(
+            &self.ck,
+            &[0]u8{},
+            2,
+            &tmp_k1,
+            &tmp_k2,
+            &out2,
+        );
+
+        var cs1 = CipherState.from_key(tmp_k1);
+        var cs2 = CipherState.from_key(tmp_k2);
+
+        return [_]CipherState{ cs1, cs2 };
+    }
+
+    pub fn clear(self: *SymmetricState) void {
+        self.cs.k = EMPTY_KEY;
+        self.ck = EMPTY_KEY;
+    }
+};
+
+// During the handshake phase each party has a single CipherState but during
+// the transport phase each party has two CipherState objects: one for sending
+// and one for receiving.
+pub const CipherState = struct {
+    // A cipher key of 32 bytes (which maybe empty).
+    k: [KEY_LEN]u8,
+
+    // An 8-byte unsigned integer nonce.
+    n: u64,
+
+    pub fn init() CipherState {
+        return .{
+            .k = EMPTY_KEY,
+            .n = 0,
+        };
+    }
+
+    pub fn from_key(key: [KEY_LEN]u8) CipherState {
+        return .{
+            .k = key,
+            .n = 0,
+        };
+    }
+
+    pub fn encrypt_with_ad(
+        self: *CipherState,
+        ad: []const u8,
+        msg: []u8,
+        mac: *[MAC_LEN]u8,
+    ) void {
+        if (!mem.eql(u8, &self.k, &EMPTY_KEY)) {
+            encrypt(msg, mac, ad, self.n, self.k);
+            self.n += 1;
+        }
+    }
+
+    pub fn decrypt_with_ad(
+        self: *CipherState,
+        ad: []const u8,
+        msg: []u8,
+        mac: [MAC_LEN]u8,
+    ) !void {
+        if (!mem.eql(u8, &self.k, &EMPTY_KEY)) {
+            try decrypt(msg, mac, ad, self.n, self.k);
+            self.n += 1;
+        }
+    }
+
+    pub fn write_msg_regular(self: *CipherState, msg: []u8) void {
+        var in_out = msg[0 .. msg.len - MAC_LEN];
+        var mac = msg[msg.len - MAC_LEN ..];
+        var tmp_mac = [_]u8{0} ** MAC_LEN;
+
+        var zerolen = [0]u8{};
+        self.encrypt_with_ad(&zerolen, in_out, &tmp_mac);
+        mem.copy(u8, mac, &tmp_mac);
+    }
+
+    pub fn read_msg_regular(self: *CipherState, msg: []u8) !void {
+        var in_out = msg[0 .. msg.len - MAC_LEN];
+        var mac = msg[msg.len - MAC_LEN ..];
+
+        var tmp_mac = [_]u8{0} ** MAC_LEN;
+        mem.copy(u8, &tmp_mac, mac);
+
+        var zerolen = [_]u8{};
+        try self.decrypt_with_ad(&zerolen, in_out, tmp_mac);
+    }
+};
+
+fn encrypt(
+    output: []u8,
+    mac: *[MAC_LEN]u8,
+    ad: []const u8,
+    nonce: u64,
+    k: [KEY_LEN]u8,
+) void {
+    ChaCha20Poly1305.encrypt(output, mac, output, ad, nonce_to_bytes(nonce), k);
+}
+
+fn decrypt(
+    output: []u8,
+    mac: [MAC_LEN]u8,
+    ad: []const u8,
+    nonce: u64,
+    k: [KEY_LEN]u8,
+) !void {
+    try ChaCha20Poly1305.decrypt(output, output, mac, ad, nonce_to_bytes(nonce), k);
+}
+
+fn nonce_to_bytes(nonce: u64) [12]u8 {
```

Comment:
std.mem.writeIntLittle(u64, &buffer[4..][0..8].*, nonce); can be used instead of this function.

Reply:
A u96 could technically be used instead, and you'd widen the nonce being a u64 to a u96 by doing @intCast(u96, nonce).